http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyRequestLifecycle.java

```diff
@@ -0,0 +1,283 @@
+            return ExecutionFlow.error(decoderResult.cause());
+        }
+        return super.fulfillArguments(routeMatch).flatMap(this::waitForBody);
+    }
+
+    /**
+     * If necessary (e.g. when there's a {@link Body} parameter), wait for the body to come in.
+     * This method also sometimes fulfills more controller parameters with form data.
+     */
+    private ExecutionFlow<RouteMatch<?>> waitForBody(RouteMatch<?> routeMatch) {
+        if (!shouldReadBody(routeMatch)) {
+            ctx.read();
+            return ExecutionFlow.just(routeMatch);
+        }
+        BaseRouteCompleter completer = nettyRequest.isFormOrMultipartData() ?
+            new FormRouteCompleter(new NettyStreamingFileUpload.Factory(rib.serverConfiguration.getMultipart(), rib.getIoExecutor()), rib.conversionService, nettyRequest, routeMatch) :
+            new BaseRouteCompleter(nettyRequest, routeMatch);
+        HttpContentProcessor processor = rib.httpContentProcessorResolver.resolve(nettyRequest, routeMatch);
+        StreamingDataSubscriber pr = new StreamingDataSubscriber(completer, processor);
+        ((StreamedHttpRequest) nettyRequest.getNativeRequest()).subscribe(pr);
+        return CompletableFutureExecutionFlow.just(pr.completion);
+    }
+
+    void handleException(Throwable cause) {
+        onError(cause).onComplete((response, throwable) -> rib.writeResponse(ctx, nettyRequest, response, throwable));
+    }
+
+    private boolean shouldReadBody(RouteMatch<?> routeMatch) {
+        if (!HttpMethod.permitsRequestBody(request().getMethod())) {
+            return false;
+        }
+        if (!(nettyRequest.getNativeRequest() instanceof StreamedHttpRequest)) {
+            // Illegal state: The request body is required, so at this point we must have a StreamedHttpRequest
+            return false;
+        }
+        if (routeMatch instanceof MethodBasedRouteMatch<?, ?> methodBasedRouteMatch) {
+            if (Arrays.stream(methodBasedRouteMatch.getArguments()).anyMatch(argument -> MultipartBody.class.equals(argument.getType()))) {
+                // MultipartBody will subscribe to the request body in MultipartBodyArgumentBinder
+                return false;
+            }
+            if (Arrays.stream(methodBasedRouteMatch.getArguments()).anyMatch(argument -> HttpRequest.class.equals(argument.getType()))) {
+                // HttpRequest argument in the method
+                return true;
+            }
+        }
+        Optional<Argument<?>> bodyArgument = routeMatch.getBodyArgument()
+            .filter(argument -> argument.getAnnotationMetadata().hasAnnotation(Body.class));
+        if (bodyArgument.isPresent() && !routeMatch.isSatisfied(bodyArgument.get().getName())) {
+            // Body argument in the method
+            return true;
+        }
+        // Might be some body parts
+        return !routeMatch.isExecutable();
+    }
+
+    private static class StreamingDataSubscriber implements Subscriber<ByteBufHolder> {
```

Comment:
I think this can also be replaced with Reactor's many sink with backpressure buffer.

Reply:
i don't think it's necessary, this one is simple enough.